src/test/unit/ntypes/ValueMapTest.cpp

```diff
@@ -0,0 +1,116 @@
+/* ---------------------------------------------------------------------
+ * HTM Community Edition of NuPIC
+ * Copyright (C) 2013, Numenta, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * Implementation of YAML tests
+ */
+
+#include "gtest/gtest.h"
+#include <htm/engine/Network.hpp>
+#include <htm/engine/Spec.hpp>
+#include <htm/ntypes/ValueMap.hpp>
+
+namespace testing { 
+    
+using namespace htm;
+
+TEST(ValueMapTest, IntScalerTest) {
+  ValueMap vm;
+
+  const char *s1 = "10";
+  vm.parse(s1);
+  EXPECT_TRUE(vm.isScalar());
+  UInt32 u = vm.as<UInt32>();
```

Comment:
> do we need to keep the value map with separate parse, "as" steps, or could we use just vm.as<T>("str") ?

I don't follow the question.   Are you proposing the argument to the as( arg ) be a string to be parsed?  I guess it could.  This would only be useful for things like default values from the Spec stored as Yaml or JSON strings.  But perhaps it might have use elsewhere.

The test is checking the ability of the root of the tree to be a scalar.  Normally it is a map or a sequence but could be any of the three or empty.   If you know ahead of time that a string contains a scalar then you don't need to parse...just use std:: conversion functions.  But if you don't know then you will need to check if the parsed node is a scalar before using the as( ).

There are lots of tests yet to be added.

Reply:
Actually, the reason I started with tests of the root being a scalar is that RapidYaml required the root node to be a map or sequence. If you parsed a string that was really a scalar it would create a sequence and add the scalar as the first node. So you could not tell the difference between "123" and "[123]". Scalar operations on the root would fail.  So I had to find a work-around.

Yaml-cpp allowed the root to be scalar.  libyaml does not have a node, you have to create it yourself...which is what our Value class is.